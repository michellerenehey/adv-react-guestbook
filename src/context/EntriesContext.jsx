import { createContext, useContext, useState } from 'react';

const EntriesContext = createContext();

function EntriesProvider({ children }) {
  const [entries, setEntries] = useState([]);

  return (
    <EntriesContext.Provider value={{ entries, setEntries }}>{children}</EntriesContext.Provider>
  );
}

function useEntries() {
  const context = useContext(EntriesContext);
  if (context === undefined) {
    throw new Error('useEntries is not defined, it needs to be in an EntriesProvider');
  }
  return context;
}

export { useEntries, EntriesProvider };
