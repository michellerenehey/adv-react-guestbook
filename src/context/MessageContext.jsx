import { createContext, useContext, useState } from 'react';

const MessageContext = createContext();

function MessageProvider({ children }) {
  const [messages, setMessages] = useState([]);

  return (
    <MessageContext.Provider value={{ messages, setMessages }}>{children}</MessageContext.Provider>
  );
}

function useMessages() {
  const context = useContext(MessageContext);
  if (context === undefined) {
    throw new Error('useMessages is not defined, it needs to be in a MessageProvider');
  }
  return context;
}

export { useMessages, MessageProvider };
