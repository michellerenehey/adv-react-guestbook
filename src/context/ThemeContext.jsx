import { createContext, useContext, useState, useEffect } from 'react';

const ThemeContext = createContext();

function ThemeProvider({ children }) {
  const [theme, setTheme] = useState('light');

  const toggleTheme = () => {
    if (theme !== 'dark') {
      localStorage.setItem('theme', 'dark');
      setTheme('dark');
    } else {
      localStorage.setItem('theme', 'light');
      setTheme('light');
    }
  };

  useEffect(() => {
    const localTheme = localStorage.getItem('theme');
    if (localTheme) {
      setTheme(localTheme);
    }
  }, []);

  return <ThemeContext.Provider value={{ theme, toggleTheme }}>{children}</ThemeContext.Provider>;
}

function useTheme() {
  const context = useContext(ThemeContext);
  if (context === undefined) {
    throw new Error('useTheme must be wrapped in a ThemeProvider');
  }
  return context;
}

export { useTheme, ThemeProvider };
