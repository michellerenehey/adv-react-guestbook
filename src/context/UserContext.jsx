import { createContext, useContext, useEffect, useState } from 'react';
import { getNames } from '../services/entries';

const UserContext = createContext();

function UserProvider({ children }) {
  const [user, setUser] = useState('');
  const [userArray, setUserArray] = useState([]);

  useEffect(() => {
    const fetchData = async () => {
      const data = await getNames();
      setUserArray(data);
    };
    fetchData();
  }, []);

  return (
    <UserContext.Provider value={{ user, setUser, userArray, setUserArray }}>
      {children}
    </UserContext.Provider>
  );
}

const useUser = () => {
  const context = useContext(UserContext);

  if (context === undefined) {
    throw new Error('useUser is not defined, it must be defined within the UserProvider');
  }

  const { user, setUser } = context;
  return { user, setUser };
};

const useUserArray = () => {
  const context = useContext(UserContext);

  if (context === undefined) {
    throw new Error('useUserArray is not defined, it must be defined within the UserProvider');
  }

  const { userArray, setUserArray } = context;
  return { userArray, setUserArray };
};

export { UserProvider, useUser, useUserArray };
