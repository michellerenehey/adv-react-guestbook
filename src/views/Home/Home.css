.Home {
  display: flex;
  flex-direction: column;
  align-items: center;
  font-family: 'Text Me One', sans-serif;
}

.book-outer {
  border: 1px solid black;
  padding: 50px;
  display: flex;
  flex-direction: column;
  align-items: start;
  margin: 20px;
  font-size: 25px;
}

.submit-button {
  text-align: center;
}

button {
  background-color: #073b4c;
  color: white;
  font-family: 'Open Sans Condensed', sans-serif;
  width: 75px;
  padding: 10px;
  margin: 10px;
}

.book-name {
  margin: 10px;
}

h3 {
  color: #073b4c;
  font-family: 'Sacramento', cursive;
  font-size: 8vh;
}

.book-header {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.signature {
  height: 10vh;
  margin: 10px;
}
