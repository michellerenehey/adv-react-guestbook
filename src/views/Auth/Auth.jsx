import './Auth.css';
import { useState } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import { useUser } from '../../context/UserContext';

export default function Auth() {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState(null);
  const { setUser } = useUser();
  const location = useLocation();
  const history = useHistory();

  const handleLogin = (e) => {
    e.preventDefault();
    if (
      username === process.env.REACT_APP_AUTH_USERNAME &&
      password === process.env.REACT_APP_AUTH_PASSWORD
    ) {
      setUser({ username });
      const { from } = location.state || { from: { pathname: '/' } };
      history.replace(from.pathname);
    } else {
      setError('Wrong credentials, try again!');
    }
  };
  return (
    <div className="Auth">
      <form>
        <h1>Login</h1>
        <input
          type="email"
          value={username}
          placeholder="enter email"
          onChange={(e) => setUsername(e.target.value)}
        />
        <input
          type="text"
          value={password}
          placeholder="enter password"
          onChange={(e) => setPassword(e.target.value)}
        />
        <button onClick={handleLogin}>Submit</button>
      </form>
      {error && <p>{error}</p>}
    </div>
  );
}
