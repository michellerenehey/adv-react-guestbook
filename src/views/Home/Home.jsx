import './Home.css';
import { useUser, useUserArray } from '../../context/UserContext';
import { useMessages } from '../../context/MessageContext';
import { useState } from 'react';
import Guestbook from '../../components/Guestbook/Guestbook';
import signature from '../../assets/signature.png';

export default function Home() {
  const [newMessage, setNewMessage] = useState('');
  const [name, setName] = useState('');
  const { user, setUser } = useUser();
  const { messages, setMessages } = useMessages();
  const { userArray, setUserArray } = useUserArray();

  // add loading state?
  // useEffect(() => {
  //   if (user.name) {
  //     setLoading(false)
  //   } else {
  //     setLoading(true)
  //   }
  // }, [user])

  const handleSubmit = async (e) => {
    e.preventDefault();
    setUser(name);
    let today = new Date();
    setMessages([...messages, { name, note: newMessage, id: today.getSeconds() }]);
    setNewMessage('');
    setUserArray([...userArray, { name: name, id: today.getSeconds() }]);
  };

  return (
    <div className="Home">
      <div className="book-header">
        <img className="signature" src={signature} alt="icon of pen" />
        <h3>Will you sign my yearbook?</h3>
        <img className="signature" src={signature} alt="icon of pen" />
      </div>
      <div className="book-outer">
        <div className="book-inner">
          <form onSubmit={handleSubmit}>
            {!user && (
              <div className="book-name">
                <label>Enter your name: </label>
                <input
                  value={name}
                  placeholder="...name..."
                  onChange={(e) => setName(e.target.value)}
                />
              </div>
            )}
            <div className="book-message">
              <label>Enter a message: </label>
              <textarea
                value={newMessage}
                placeholder="...message..."
                onChange={(e) => setNewMessage(e.target.value)}
              />
            </div>
            <div className="submit-button">
              <button type="submit">Submit</button>
              {user && (
                <button
                  onClick={() => {
                    setUser(''), setName('');
                  }}
                >
                  Not {user}?
                </button>
              )}
            </div>
          </form>
        </div>
      </div>
      <Guestbook />
    </div>
  );
}
