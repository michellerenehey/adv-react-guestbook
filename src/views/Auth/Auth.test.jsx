import {render} from '@testing-library/react'; 
import {MemoryRouter, Route, Switch} from 'react-router-dom'
import { UserProvider } from '../../context/UserContext';
import { EntriesProvider } from '../../context/EntriesContext';
import { ThemeProvider } from '../../context/ThemeContext';
import Auth from './Auth.jsx'; 
import Home from '../Home/Home.jsx'; 

test('user can login', async () => {
    const {container} = render(
    <UserProvider>
      <EntriesProvider>
        <ThemeProvider>
            <MemoryRouter>
                <Switch>
                    <Route exact path="/login">
                        <Auth />
                    </Route>
                    <Route exact path="/">
                        <Home />
                    </Route>
                </Switch>
            </MemoryRouter>
        </ThemeProvider>
      </EntriesProvider>
    </UserProvider>
    ); 

    expect(container).toMatchSnapshot();
})