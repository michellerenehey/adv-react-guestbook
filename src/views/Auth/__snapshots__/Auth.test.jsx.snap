// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user can login 1`] = `
<div>
  <div
    class="Home"
  >
    <div
      class="book-header"
    >
      <img
        alt="icon of pen"
        class="signature"
        src="signature.png"
      />
      <h3>
        Will you sign my yearbook?
      </h3>
      <img
        alt="icon of pen"
        class="signature"
        src="signature.png"
      />
    </div>
    <div
      class="book-outer"
    >
      <div
        class="book-inner"
      >
        <form>
          <div
            class="book-message"
          >
            <label>
              Enter a message: 
            </label>
            <textarea
              placeholder="...message..."
            />
          </div>
          <div
            class="submit-button"
          >
            <button
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      class="Guestbook"
    />
  </div>
</div>
`;
